### ISSUE TITLE
Deployment failure when using the contains operator with string properties in apply rules

### ISSUE THREAD

**Issue by lippserd** [Icinga/icingaweb2-module-director#809]
When creating a service apply rule using the contains operator with host.name for example, the generated configuration is invalid because the Director renders this as `"filter" in host.name` which is not yet supported: Icinga/icinga2/#5018.

It would be optimal, if the Director offers the appropriate operators based on the type of the property. But I think it would be sufficient for the moment to document that the contains operator is only valid for array properties.

@Thomas-Gelf What are your thoughts about this?

**Comment by Thomas-Gelf** [comment]
Added my comments over there. This is at least a documentation/GUI issue, "contains" should not be shown when dealing with non-array values. Just, this is sometimes not as clear as one might think - as the same variable might be filled with either strings or arrays depending on who/where you are.

The correct way to accomplish wild-card matches in the Director is by using the = (equals) operator combined with wild-cards (using the asterisk `*`).` host.name = *filter*` renders to `where match("*filter*", host.name)*`.

However, I fully understand the problem that "contains" might be misleading especially to people new to the Director (read: those who didn't use it before 1.3.0). This should somehow be addressed.

**Comment by pillbug22** [comment]
+1
This fixed a frustrating issue I was having trying to create Service Groups - just needing to use "=" with wildcards instead of the "contains" filter.

**Comment by lippserd** [comment]
Hi Tom,

We decided to not support the in operator for strings at the moment. It would be great if you could add a hint somewhere that the contains operator will not work for strings.

Cheers,
Eric

**Comment by Thomas-Gelf** [comment]
@lippserd: yep, there should be a related issue. Just didn't know where to place the hint :laughing: Good decision to not allow this in Icinga, also IMO it would make no sense.

**Comment by carraroj** [comment]
ref/IP/14611

**Comment by Slevinox** [comment]
Hello,
Just to understand is it wanted in Director implementation "contains" is replaced by "in" ?
I mean instead of using the contains method $string_var$.contains($string$).

**Comment by bitboy85** [comment]
This is confusing that you have to use = with wildcard. Usually wildcards only work with LIKE.

PS: I guess theres also a director gui issue with IN. if you have selected it you cannot switch back to another operator.
